### ISSUE TITLE
radio control use align center and overflows + can not set static parent name 

### ISSUE THREAD

**Issue by trackme518** [KrabCode/LazyGui#264]
Hi,
when using:
`gui.radio("mode", new String[]{"super long option that does not fit inside the windows", "circle"});`
the option will overflow from the control window. 
### Proposed solution: 
- Align left not center and then crop rest of the string that overflows. 

Another problem is that the folder name will be renamed based on the selected option - this is a feature but it would be nice to provide an option to disable this behavior and keep the folder name static as it was declared - perhaps with new options toggle?

**Comment by KrabCode** [comment]
Hey, thanks for the new issue!

Overflows are a problem in general, not only with this but also with float values and long folder names... I'll have to take a more comprehensive look at them sometime.

But I'd like to say that the folder name itself does not change, it still says "mode" - I think you'd see that if you just modify your example to have a shorter first option. Like this: `gui.radio("mode", new String[]{"square", "circle"});`

**Comment by trackme518** [comment]
Hi,
I was just scanning the source code and somehow it seems already implemented somewhere?
```
   protected void drawLeftText(PGraphics pg, String text){
        fillForegroundBasedOnMouseOver(pg);
        String trimmedText = FontStore.getSubstringFromStartToFit(pg, text, size.x - FontStore.textMarginX);
        pg.textAlign(LEFT, CENTER);
        pg.text(trimmedText, FontStore.textMarginX, LayoutStore.cell - FontStore.textMarginY);
    }

    protected void drawRightText(PGraphics pg, String text, boolean fillBackground) {
        if(fillBackground){
            float w = pg.textWidth(text) + FontStore.textMarginX * 2;
            drawRightBackdrop(pg, w);
        }
        pg.textAlign(RIGHT, CENTER);
        pg.text(text,
                size.x - FontStore.textMarginX,
                size.y - FontStore.textMarginY
        );
    }
```
It seems that you are already doing some text trimming in AbstractNode.java line 121. 

And in FontStore.java line 102:

```
    public static String getSubstringFromStartToFit(PGraphics pg, String text, float availableWidth) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            char character = text.charAt(i);
            float textWidthAfterNewChar = pg.textWidth(result.toString() + character);
            if (textWidthAfterNewChar >= availableWidth) {
                break;
            }
            result.append(character);
        }
        return result.toString();
    }
```
But it seems unnecessary complicated - you can use:
`textWidth(String text);`
See: https://processing.org/reference/textWidth_.html

**Comment by trackme518** [comment]
I also did a test sketch and it seems to work:

```
String longtext = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
String trimmedtext;
float avaliableWidth = 200.0;

void setup() {
  size(640, 480, P2D);
  textSize(24);
  trimmedtext = getSubstringFromStartToFit(longtext, avaliableWidth);
}

void draw() {
  background(0);
  fill(255);
  stroke(255);
  text(trimmedtext, 0,30);
  line(avaliableWidth,0,avaliableWidth,height);
}

public String getSubstringFromStartToFit( String text, float availableWidth) {
  StringBuilder result = new StringBuilder();
  for (int i = 0; i < text.length(); i++) {
    char character = text.charAt(i);
    float textWidthAfterNewChar = textWidth(result.toString() + character);
    if (textWidthAfterNewChar >= availableWidth) {
      break;
    }
    result.append(character);
  }
  return result.toString();
}

```

So I wonder if the problem is in the way you are setting the font or PGraphics canvas?

**Comment by KrabCode** [comment]
I don't really have time to delve into this right now but feel free to submit a pull request if you identify the problem.

And how about the folder renaming? Could you please clarify what you mean? Ideally provide a test sketch?

**Comment by trackme518** [comment]
By folder renaming I mean that when you are using gui.radio the name of that folder in parent window will get renamed by selected option (mode: selectedoption) but sometimes I don't want that - I would prefer to have it stay just "mode" in parent window. And that could be a selection in theme options.

**Comment by KrabCode** [comment]
Well it's not renaming the folder then... it just displays the current value.

Folder name is on the left, value string is on the right.

**Comment by KrabCode** [comment]
I guess the value display in the parent row could be hidden, sure. I don't see the benefit personally, but it can be an option in the constructor settings.

**Comment by trackme518** [comment]
Hi,
I was able to reproduce the issue - it might be caused by PGraphics not taking into account font size?

```
String longtext = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
String trimmedtext;
float avaliableWidth = 200.0;
PFont font;
PGraphics pg;

void setup() {
  size(960, 480, P2D);
  pg = createGraphics(width/2, height, P2D);
  font = createFont("JetBrains Mono", 24);
  textFont(font); //set font for default render buffer
  pg.textFont(font);//set font for PGrahics buffer
  //textSize(24);
  trimmedtext = getSubstringFromStartToFit(longtext, avaliableWidth);
}

void draw() {
  background(0);
  fill(255);
  stroke(255);
  text(trimmedtext, width/2 + 0, 30);
  line( width/2 + avaliableWidth, 0, width/2 + avaliableWidth, height);

  pg.beginDraw();
  pg.background(0);
  pg.fill(255);
  pg.stroke(255);
  pg.text(trimmedtext, 0, 30);
  pg.line(avaliableWidth, 0, avaliableWidth, height);
  pg.endDraw();
  image(pg, 0, 0);
}

public String getSubstringFromStartToFit( String text, float availableWidth) {
  StringBuilder result = new StringBuilder();
  for (int i = 0; i < text.length(); i++) {
    char character = text.charAt(i);
    float textWidthAfterNewChar = textWidth(result.toString() + character);
    if (textWidthAfterNewChar >= availableWidth) {
      break;
    }
    result.append(character);
  }
  return result.toString();
}

```
![notCorrectPGrahics](https://github.com/KrabCode/LazyGui/assets/50262885/08764551-91c3-47e6-a282-f9b32ecc9010)

**Comment by trackme518** [comment]
ABstractNode.java line 128:

```
    protected void drawRightText(PGraphics pg, String text, boolean fillBackground) {
        if(fillBackground){
            float w = pg.textWidth(text) + FontStore.textMarginX * 2;
            drawRightBackdrop(pg, w);
        }
        pg.textAlign(RIGHT, CENTER);
        pg.text(text,
                size.x - FontStore.textMarginX,
                size.y - FontStore.textMarginY
        );
    }
```
You are not calling `String trimmedText = FontStore.getSubstringFromStartToFit(pg, text, size.x - FontStore.textMarginX);` instead you are drawing the text directly in its full lenght: `pg.text(text,....`

**Comment by trackme518** [comment]
I have created a pull request - I did not test it since I didn't setup the build for the library in  Eclipse - please verify first and create a new build we can download. Thanks!

**Comment by KrabCode** [comment]
It looks great, I cleaned it up a bit and made it buildable and extended the same logic to all nodes that display text on the right of their rows. It doesn't overflow the left text and it doesn't go beyond the window either. I'll test it some more soon and see if it's ready for pushing to master and deploying to processing but you can have the jar already to test it yourself.

![image](https://github.com/KrabCode/LazyGui/assets/25923016/c015140d-0668-4e23-9505-14983b246473)
![image](https://github.com/KrabCode/LazyGui/assets/25923016/e50d7f41-3604-48df-ab4b-779ba93b56df)
![image](https://github.com/KrabCode/LazyGui/assets/25923016/f1d26f29-2ba2-4333-a464-026664036b75)

**Comment by KrabCode** [comment]
you can find the test release here: https://github.com/KrabCode/LazyGui/releases/tag/test-264
drag & drop the jar into PDE or use it in Eclipse and you should be able to see the changes

**Comment by trackme518** [comment]
OK, I see the problem with this - the text get cropped correctly but it will get cropped from the left...to me it make more sense when it would be cropped from right - see the attached image ("MIDI device" is the label and "ble Synth" is the value text cropped from left). So really there should be "getSubstringFromStartToFit" NOT the "getSubstringFromEndToFit" :-/ sorry :-D

Also I have hinted on that you can use text(string,x,y,width,height) to crop text using processing built in function - maybe be it would be more efficient?

![leftCropped](https://github.com/KrabCode/LazyGui/assets/50262885/a64310e4-e46d-46fc-b643-d3bfd7133d8b)

On other note - and I should probably start new Issue for this - I would love to have ability to delete the node - currently I can only .hide() it so it is not visible but that is not memory efficient - I can live with that but it would be nice. I am spawning new GUI controllers as new clients connect and when they disconnect or change their parameters I am hiding them and creating new ones... but maybe it is possible already?

Another stuff I will look into later is porting this to Android - I believe it should be possible - I saw an issue on that, but that will come later, just letting you know.

**Comment by KrabCode** [comment]
I'm using IntelliJ IDEA, there's a guide in the readme. I'll have to improve the build process someday to use gradle and not rely on manual library hunting. I'll make an issue for that if I don't find one already. There was one contributor who added maven for the same purpose but I am not familiar with that as much as gradle.

Please make a new issue for the node deletion. 

Also please make another issue for the optional right value text hiding.

Yay for android, I just fear we'll have to drop all the shaders for it to work. And rework the mouse / keyboard interaction and scaling and saving and... lots of stuff to do. I'd definitely recommend developing that in a separate git branch.

**Comment by KrabCode** [comment]
I changed it to show the start rather than the end, feel free to test it with the new jar. I'll probably release this in a few days, I'll need some time to test it...

https://github.com/KrabCode/LazyGui/releases/tag/test-264

**Comment by KrabCode** [comment]
I don't like how resizing the window makes the text movement switch between smooth and sudden when one whole letter gets trimmed, I'd like to fine-tune that if possible.

**Comment by KrabCode** [comment]
I released it along with #268 - you can find v1.3.2 in the processing editor and I'm closing the issue for now.
